src/diffusers/models/noise_augmentor.py

```diff
@@ -0,0 +1,167 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from typing import List, Optional, Union
+
+import numpy as np
+import torch
+from torch import nn
+
+from ..configuration_utils import ConfigMixin, register_to_config
+from ..utils import randn_tensor
+from .embeddings import Timesteps
+from .modeling_utils import ModelMixin
+
+
+# Copied from diffusers.schedulers.scheduling_ddpm.betas_for_alpha_bar
+def betas_for_alpha_bar(num_diffusion_timesteps, max_beta=0.999) -> torch.Tensor:
+    """
+    Create a beta schedule that discretizes the given alpha_t_bar function, which defines the cumulative product of
+    (1-beta) over time from t = [0,1].
+
+    Contains a function alpha_bar that takes an argument t and transforms it to the cumulative product of (1-beta) up
+    to that part of the diffusion process.
+
+
+    Args:
+        num_diffusion_timesteps (`int`): the number of betas to produce.
+        max_beta (`float`): the maximum beta to use; use values lower than 1 to
+                     prevent singularities.
+
+    Returns:
+        betas (`np.ndarray`): the betas used by the scheduler to step the model outputs
+    """
+
+    def alpha_bar(time_step):
+        return math.cos((time_step + 0.008) / 1.008 * math.pi / 2) ** 2
+
+    betas = []
+    for i in range(num_diffusion_timesteps):
+        t1 = i / num_diffusion_timesteps
+        t2 = (i + 1) / num_diffusion_timesteps
+        betas.append(min(1 - alpha_bar(t2) / alpha_bar(t1), max_beta))
+    return torch.tensor(betas, dtype=torch.float32)
+
+
+class NoiseAugmentor(ModelMixin, ConfigMixin):
+    """
+    NoiseAugmentor is used in the context of stable unclip to add noise to the image embeddings. The amount of noise is
+    controlled by a `noise_level` input. A higher `noise_level` increases the variance in the final un-noised images.
+
+    The noise is applied in two ways
+    1. A noise schedule is applied directly to the embeddings
+    2. A vector of sinusoidal time embeddings are appended to the output.
+
+    In both cases, the amount of noise is controlled by the same `noise_level`.
+
+    The embeddings are normalized before the noise is applied and un-normalized after the noise is applied, so the
+    layer requires parameters for the mean and standard deviation of the embeddings.
+    """
+
+    @register_to_config
+    def __init__(
+        self,
+        embedding_dim: int = 768,
+        max_noise_level: int = 1000,
+        beta_start: float = 0.0001,
+        beta_end: float = 0.02,
+        beta_schedule: str = "linear",
+        trained_betas: Optional[Union[np.ndarray, List[float]]] = None,
+        set_alpha_to_one: bool = True,
+    ):
+        super().__init__()
+        self.max_noise_level = max_noise_level
+
+        self.mean = nn.Parameter(torch.zeros(1, embedding_dim))
```

Comment:
Are those parameters learned?

Reply:
E.g. are any of the parameters of this module learned through training or are they just set and fixed?

Reply:
They're set based off of which CLIP model is used for the image embeddings. 

For the openai clip, they're set to the calculated mean and stddev. For the openclip, they're left as 0 and 1 respectively.

Reply:
resolving with switch to StableUnCLIPImageNormalizer